capreolus/searcher/special.py

```diff
@@ -0,0 +1,196 @@
+            return "https://msmarco.blob.core.windows.net/msmarcoranking/qidpidtriples.train.full.2.tsv.gz"
+
+        if tripleversion == "small":
+            return "https://drive.google.com/uc?id=1LCQ-85fx61_5gQgljyok8olf6GadZUeP"
+
+        raise ValueError("Unknown version for triplet large" % self.config["tripleversion"])
+
+    def download_and_prepare_train_set(self, tmp_dir):
+        tmp_dir.mkdir(exist_ok=True, parents=True)
+        triple_version = self.config["tripleversion"]
+
+        url = self.get_url()
+        if triple_version.startswith("large"):
+            extract_file_name = self.get_fn_from_url(url)
+            extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+            triple_fn = extract_dir / extract_file_name
+        elif triple_version == "small":
+            triple_fn = tmp_dir / "triples.train.small.idversion.tsv"
+            if not os.path.exists(triple_fn):
+                gdown.download(url, triple_fn.as_posix(), quiet=False)
+        else:
+            raise ValueError(f"Unknown version for triplet: {triple_version}")
+
+        return self.convert_to_trec_runs(triple_fn, style="triple")
+
+
+@Searcher.register
+class MsmarcoPsg(Searcher, MsmarcoPsgSearcherMixin):
+    module_name = "msmarcopsg"
+    dependencies = [Dependency(key="benchmark", module="benchmark", name="msmarcopsg")]
+    config_spec = [
+        ConfigOption("tripleversion", "small", "version of triplet.qid file, small, large.v1 or large.v2"),
+    ]
+
+    def _query_from_file(self, topicsfn, output_path, cfg):
+        """only query results in dev and test set are saved"""
+        final_runfn = output_path / "searcher"
+        final_donefn = output_path / "done"
+        if os.path.exists(final_donefn):
+            return output_path
+
+        tmp_dir = self.get_cache_path() / "tmp"
+        tmp_dir.mkdir(exist_ok=True, parents=True)
+        output_path.mkdir(exist_ok=True, parents=True)
+
+        # train
+        train_run = self.download_and_prepare_train_set(tmp_dir=tmp_dir)
+        self.write_trec_run(preds=train_run, outfn=final_runfn, mode="wt")
+
+        # dev and test
+        dev_test_urls = [
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.dev.tar.gz",
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.eval.tar.gz",
+        ]
+        runs = {}
+        for url in dev_test_urls:
+            extract_file_name = self.get_fn_from_url(url)
+            extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+            runs.update(self.convert_to_trec_runs(extract_dir / extract_file_name, style="eval"))
+        self.write_trec_run(preds=runs, outfn=final_runfn, mode="a")
+
+        with open(final_donefn, "wt") as f:
+            print("done", file=f)
+        return output_path
+
+
+@Searcher.register
+class MsmarcoPsgBm25(BM25, MsmarcoPsgSearcherMixin):
```

Comment:
Is the purpose of this searcher to exclude train queries when searching? Can you add a short docstring about what it's for?